### ISSUE TITLE
Further PyTorch API alignment?

### ISSUE THREAD

**Issue by dsyme** [DiffSharp/DiffSharp#128]
If we look at this PyTorch code:
```python
input = torch.tensor([1.0,2.0,3.0], requires_grad=True) 
loss = input.mean() 
loss.backward(torch.tensor(3.0)) 
input.grad 
```
The DiffSharp code is as follows:
```fsharp
    let input = dsharp.tensor([1.0,2.0,3.0]).reverseDiff() 
    let loss = input.mean() 
    loss.reverse(dsharp.tensor(30.)) 
    input.derivative
```
Questions

* We define both `tensor.backward(inp)` and `tensor.reverse(inp)` - do we really want both?  I guess it doesn't really matter.  Do we have a preference which gets used in examples?

* Should we rename `tensor.derivative` to `tensor.gradient` or `tensor.grad`?   My impression is people are now generally more comfortable with the term "gradient" in the context of actual machine learning.  Maybe it doesn't matter though and we should just have both.

There are also lots of other places where we could align the API much more closely to PyTorch.  I think on the whole it would be worth doing, even if it means using underscores in names.  (We can always consider a .NET style API separately but since we're most the way there already it feels like we should just go the whole way)

**Comment by gbaydin** [comment]
These things about the differentiation API are still subject to change and exploration. I will need more time to explore the options and think about it.

My two quick comments are:
- DiffSharp is not a PyTorch wrapper. It's just that the naming and API design in DiffSharp are influenced PyTorch, to support the latest trends in ML workflow that would be familiar for many people. There are parts of the API that are about functional differentiable programming (the `grad`, `jacobianv`, `hessianv` etc.) that is inherited from the original DiffSharp, and there is a new low-level differentiation API (`backward`, `.derivative`, `.primal`, etc.) that are more PyTorch-like and are based around methods on tensors. We could call these function-oriented and Tensor-oriented. Putting these together into one system is an interesting design challenge. I'm still in the process of designing this.

- I'm very strongly against the `grad` name to refer to a general derivative, that is something PyTorch ended up inheriting from the Python `autograd` package, and is ultimately because of the assumption of backpropagation from a scalar value (the standard use case of derivatives in ML) in order to perform gradient descent. These are not definitely not gradients, gradient-descent is not the only thing one can do with a derivative, and using `gradient` to refer to general derivative is mathematically incorrect. Furthermore in DiffSharp a `.derivative` can hold a forward mode derivative which is almost never a gradient in the mathematical sense. The rest of the ML frameworks are slowly discovering that a gradient (and backpropagation) is not the only type of derivative one could have, and adjusting their APIs accordingly (e.g., the recent introductions of `jvp`, `vjp` operations in PyTorch). Using `grad` in DiffSharp is incompatible with the rest of the library spirit and design in which we have well-defined mathematical derivative operations, and nesting of forward and reverse derivative.

The `backward` and `reverse` names are for experimentation about how the API will feel like. I think `reverse` is a better name consistent with the rest of the library.

**Comment by dsyme** [comment]
Thank you, that's a fabulous explaination of both the design principle and the specific using of "gradient"!

Would it be ok if I added these to a DESIGN_NOTES.md?  I feel it's important to record these and not just in github issues,

Apart from that we can close the issue - I might return with more questions.

**Comment by dsyme** [comment]
Closing as my questions have been answered
